geomfn: fix small bugs in ST_LineSubString

otan opened the pull request:
Release note (bug fix): Fix a bug where a substring of a linestring with
the same points would return EMPTY instead of a POINT with the repeated
point.

Release note (bug fix): Fix a bug where it is possible for a linestring
returned from ST_LineSubString to have repeated points.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66738)
<!-- Reviewable:end -->

sumeerbhola reviewed (changes requested):
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66738)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/geo/geomfn/linestring.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66738#-MctESdSCU7gziygy-aZ:-MctESdT46D-iPjzDAC-:b-fixsdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/2245b7e89c9ae32f0f0773cf634183de6e529440/pkg/geo/geomfn/linestring.go#L279)):*
> ```Go
> 	}
> 
> 	// Flat line should be return the first point.
> ```

nit: why is this being referred to as "flat line"?
___
*[pkg/geo/geomfn/linestring.go, line 317 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66738#-MctG0Se9ssOnDAx42hJ:-MctG0Se9ssOnDAx42hK:b-p8rq3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/2245b7e89c9ae32f0f0773cf634183de6e529440/pkg/geo/geomfn/linestring.go#L317)):*
> ```Go
> 		if currentLineString.Length() >= startDistance {
> 			if len(newFlatCoords) == 0 {
> 				// Add the initial point.
> ```

I'm having trouble visualizing what is happening in this algorithm.
I think what is happening here is that `currentLineString` which is composed of multiple line segments, has a last line segment that contains the point corresponding to startDistance. And the interpolation is finding that point and adding it. Then that last line segment has an end point, which we also need to add if it is not the same as the point we just added. Is that correct?
A code comment with a couple of examples would really help.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66738)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @sumeerbhola)

___
*[pkg/geo/geomfn/linestring.go, line 317 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66738#-MctG0Se9ssOnDAx42hJ:-Mcvn2ZNC0BmtS66Gu_a:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2245b7e89c9ae32f0f0773cf634183de6e529440/pkg/geo/geomfn/linestring.go#L317)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I'm having trouble visualizing what is happening in this algorithm.
I think what is happening here is that `currentLineString` which is composed of multiple line segments, has a last line segment that contains the point corresponding to startDistance. And the interpolation is finding that point and adding it. Then that last line segment has an end point, which we also need to add if it is not the same as the point we just added. Is that correct?
A code comment with a couple of examples would really help.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66738)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @rafiss, and @sumeerbhola)

___
*[pkg/geo/geomfn/linestring.go, line 340 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66738#-McyZ2fn8TE7kNo5Dugf:-McyZ2fn8TE7kNo5Dugg:bbjwyef) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad716f9815148e25ad6ba7ca07086f3309f23592/pkg/geo/geomfn/linestring.go#L340)):*
> ```Go
> 				// Otherwise we have have the first point, so include the current coordinate
> 				// in the substring.
> 				newFlatCoords = append(newFlatCoords, lineString.Coord(i)...)
> ```

nit: the two statements that are appending the last point in this segment are slightly different but semantically identical -- one uses currentLineString.Coord(i) and this uses lineString.Coord(i). Could you make them the same?

Isn't it possible to have subsequent line segments that are 0 length, so every such append should check that the coordinate being appended is not the same as the previous one? Or is that where we want to preserve the "structure" of the original line string in terms of having some 0 length line segments and not collapsing them? What does PostGIS do?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66738)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @sumeerbhola)

___
*[pkg/geo/geomfn/linestring.go, line 340 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66738#-McyZ2fn8TE7kNo5Dugf:-Md-QKql89bzQ8I7uRqL:bptl0wh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad716f9815148e25ad6ba7ca07086f3309f23592/pkg/geo/geomfn/linestring.go#L340)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: the two statements that are appending the last point in this segment are slightly different but semantically identical -- one uses currentLineString.Coord(i) and this uses lineString.Coord(i). Could you make them the same?

Isn't it possible to have subsequent line segments that are 0 length, so every such append should check that the coordinate being appended is not the same as the previous one? Or is that where we want to preserve the "structure" of the original line string in terms of having some 0 length line segments and not collapsing them? What does PostGIS do?

</blockquote></details>

Good catch, PostgreSQL deduplicates the points! updated accordingly.


<!-- Sent from Reviewable.io -->

otan committed:
geomfn: fix small bugs in ST_LineSubString

Release note (bug fix): Fix a bug where a substring of a linestring with
the same points would return EMPTY instead of a POINT with the repeated
point.

Release note (bug fix): Fix a bug where it is possible for a linestring
returned from ST_LineSubString to have repeated points.

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66738#-:-Md0-6mJ2aFA7xjrFdTI:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss and @sumeerbhola)



<!-- Sent from Reviewable.io -->

otan commented:
thanks

bors r=sumeerbhola
